feat(workflow): add e2e workflow steps to run tests on remote amc hosted on AWS [WD-26499]

tarek-hanna opened the pull request:
## Done
This PR implements additional steps in `nia-e2e.yaml` to run e2e tests using `amc connect` on a remote `amc` hosted on AWS.

### Steps: ###
 **1.  "Setup remote amc"** 
- if remote url and certificates are provided in the repo secrets, amc remote url will be added and cert trusted
- remote url (if provided) must be public and in this format `https://<remote-amc-ip>:8444`
- if remote url is not provided, the github runner host ip will be used instead
- the remote amc is set as default

**2. "Create Android Instance"**
- launches an android instance with streaming enabled
- outputs instance id for the following steps
 
**3."Connect over adb"**
- runs `amc connect` command to create a sharing link and initiate adb connection
- `amc connect` is a wrapper for `anbox-connect` and is ONLY available on `snap install anbox-cloud-appliance --channel=1.28/edge/amc-connect` or above
-  Once the adb link (IP & PORT) is generated, `adb connect 127.0.0.1:$PORT` is called to attach device 

**4."Run e2e tests"**
- execute the `./gradlew :app:connectedDemoDebugAndroidTest` command.

**5."Upload e2e test reports"**
- uploads the e2e report to the artifacts

**6."Clean up android instance"**
- deletes the android instance after completing the tests

## QA
1. Perform the following QA steps:
 - submit a PR to trigger `nia-pr.yaml` workflow in which the e2e job will perform the installation of anbox appliance, launch android instance, connect adb and run the e2e test
 - All checks should pass as shown in the screenshots below
 - An artifact "nia-e2e-test" should be available
 
2. You can repeat the steps but this time make sure before you do, that the `REMOTE_AMC_URL`, `REMOTE_SERVER_CERT`, `GH_RUNNER_CERT` and `GH_RUNNER_KEY` are added to the repo; otherwise the runner's internal ip will be used as the remote amc.
- GH_RUNNER cert should be trusted on the remote aws instance
- again follow the same steps as (# 1) but this time the instance will run on the remote amc provided
- steps should pass and produce the report artifact

## JIRA / Launchpad bug

Fixes # [Anbox Cloud Demo - Add workflow steps to run e2e tests on remote amc hosted on AWS](https://warthogs.atlassian.net/browse/WD-26499)
[Update anbox-cloud-appliance snap to 1.28/edge](https://warthogs.atlassian.net/browse/WD-29854)

## Documentation
- README includes updated steps to run in github hosted runner and attach a remote amc client

## Screenshots
**Android instance creation on remote ams in realtime with the runner**

[![Watch the video](https://github.com/user-attachments/assets/cc07f662-7eb7-4c61-b38c-bec371428dec)](https://github.com/user-attachments/assets/cc07f662-7eb7-4c61-b38c-bec371428dec)

**CI and e2e Jobs completing successfully**
<img width="1920" height="719" alt="Screenshot from 2025-10-10 19-10-17" src="https://github.com/user-attachments/assets/1ccf07e4-50a7-474f-8e83-685e8b669c68" />

**e2e test report downloadable from artifacts**
<img width="1901" height="914" alt="Screenshot from 2025-10-10 19-10-52" src="https://github.com/user-attachments/assets/1f3be009-3814-421f-9659-f71741c76919" />

**e2e test report output**
<img width="874" height="755" alt="Screenshot from 2025-10-10 19-41-05" src="https://github.com/user-attachments/assets/6bdc70ae-5068-4abc-8e34-c0715bf3a413" />

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,88 @@ jobs:
       uses: canonical/anbox-cloud-github-action@dc99b94280cfcf44a86b62aeac25ba34f14725e0
       with:
-        channel: latest/stable
+        channel: 1.28/edge/amc-connect
```
We should wait before merging this on https://github.com/canonical/ams/pull/1547 and switch to `1.28/edge`. Also we need to switch back to `latest/stable` once 1.28 is out. Do you have an item to track that part?

tarek-hanna replied:
**Backlog items created :** 
[Update anbox-cloud-appliance snap to 1.28/edge](https://warthogs.atlassian.net/browse/WD-29854)
[Update anbox-cloud-appliance snap to latest/stable](https://warthogs.atlassian.net/browse/WD-29855)

morphis replied:
The change is merged so we can update 1.28/edge now

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,88 @@ jobs:
+        fi
+
+        if [ -n "${REMOTE_CLIENT_CERT}" ]; then
+          echo "${REMOTE_CLIENT_CERT}" >> "$HOME"/snap/anbox-cloud-appliance/current/client/remote_client.crt
```
```suggestion
          echo "${REMOTE_CLIENT_CERT}" > "$HOME"/snap/anbox-cloud-appliance/current/client/remote_client.crt
```

Otherwise you're amending to a potentially existing file. Aso why not just putting the file into `$HOME`?

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,88 @@ jobs:
+
+        if [ -n "${REMOTE_CLIENT_CERT}" ]; then
+          echo "${REMOTE_CLIENT_CERT}" >> "$HOME"/snap/anbox-cloud-appliance/current/client/remote_client.crt
+          amc config trust add "$HOME"/snap/anbox-cloud-appliance/current/client/remote_client.crt
```
Lets create an identity instead of using the old certificate path


```suggestion
          echo "$REMOTE_CLIENT_CERT" | amc auth identity create tls/test-user -
```

That also lets you avoid writing the cert anywhere.

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,88 @@ jobs:
+          echo "${REMOTE_CLIENT_KEY}" >> "$HOME"/snap/anbox-cloud-appliance/current/client/remote_client.key
+        fi
+
+        amc remote add remote_amc_client "${url}" --accept-certificate
```
Instead of using `--accept-certificate` it would be better security wise if we don't blindly trust the server. We can also store the servers cert in a secret and write to `$HOME/snap/anbox-cloud-appliance/current/client/servercerts/server.crt` That will make amc recognize the certificate and trust it when presented by the server allowing us here to drop `--accept-certificate`.

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,88 @@ jobs:
+        id="$(amc launch --enable-streaming jammy:android${ANDROID_VERSION}:amd64)"
         amc wait -c status=running "$id"
-        echo "id=$id" >> "$GITHUB_OUTPUT"
+        echo "id=$id" >> "$GITHUB_OUTPUT"        
```
```suggestion
        echo "id=$id" >> "$GITHUB_OUTPUT"
```

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,88 @@ jobs:
+        echo "id=$id" >> "$GITHUB_OUTPUT"        
 
-    - name: Access android over adb
+    - name: Run e2e tests over adb
```
Shouldn't we keep independent steps for this rather than using one combined step?

tarek-hanna replied:
split the ADB step from e2e

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,88 @@ jobs:
+        port=$(grep -oP 'adb connect 127.0.0.1:\K\d+' logs/anbox-connect.log)
+                  
+        adb connect 127.0.0.1:"$port"
+        # ./gradlew :app:connectedDemoDebugAndroidTest
```
Why is this commented?

tarek-hanna replied:
uncommented

morphis commented on .github/workflows/nia-pr.yaml:
```diff
@@ -13,11 +13,11 @@ concurrency:
-    secrets: inherit
+  # nia-ci:
+  #   uses: ./.github/workflows/nia-ci.yaml
+  #   secrets: inherit
```
Why is this commented?

tarek-hanna replied:
uncommented

morphis commented on .github/workflows/nia-pr.yaml:
```diff
@@ -13,11 +13,11 @@ concurrency:
   
   nia-e2e:
-    needs: nia-ci
+    # needs: nia-ci
```
Why is this commented?

tarek-hanna replied:
uncommented

morphis commented on README.md:
```diff
@@ -48,10 +48,21 @@ export PATH=$ANDROID_HOME/platform-tools:$PATH
+**Notes**
+- Another workflow is `cache-images.yaml` to download and cache Android images, that is scheduled to run every day at 10:00 AM, or can manually be dispatched.
+
+- To add a remote amc where you launch the android image, and connect to adb for running e2e tests, you must provide the following in the repos secrets: otherwise the runner's internal ip will be used as the remote url.
```
Only saw this here but we should use correct capitalized names, e.g. `Android` rather than `android`

tarek-hanna replied:
Done

ajanon commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,88 @@ jobs:
+        url=${REMOTE_AMC_URL}
+        if [ -z "${url}" ]; then
+            ip=$(hostname -I | awk '{print $1}')
+            url="https://${ip}:8444"
```
This will take a random address if we have multiple interfaces (excluding loopback)
Why not just take the address from AMS?

If we have yq (not sure if we do here):
```suggestion
            url="$(amc config show | yq -r '.config."core.https_address"')"
```

Or, with grep:

```suggestion
            url="$(amc config show | grep -Po 'core\.https_address:\s*\K.*')"
```

ajanon commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,88 @@ jobs:
+        for i in {1..5}; do
+          grep "adb connect 127.0.0.1" logs/anbox-connect.log && break
+          sleep 5
+        done
```
You need to check at the end of the loop whether you have fully timed out or not.
Consider this example:
You grep for 'adb connect' multiple times. At the end of the last loop, you end up with i=5, then you sleep. You then exit the loop normally, and you then try to grep below the port, even though you don't have it.

You should check for the value, e.g.:

```suggestion
        for i in {1..5}; do
          grep "adb connect 127.0.0.1" logs/anbox-connect.log && break
          sleep 5
        done
        if [[ "${i}" -ge 5 ]]; then
          echo "adb connect failed"
          exit 1
        fi
```

The above doesn't really work as-is, though, as if you match in the last loop, you will still fail the loop.

I would suggest using a separate function instead, e.g.:



```suggestion
        wait_for_adb_connect() {
          for i in {1..5}; do
            if grep "adb connect 127.0.0.1" logs/anbox-connect.log; then
              return 0
            fi
            sleep 5
          done
          return 1
        }
        wait_for_adb_connect
```

IIRC, all GH scripts run with errexit, so if wait_for_adb_connect returns 1 you should exit.

Also, I think we can have IPv6-only loopback interfaces (though that would be pretty weird), so only checking for 127.0.0.1 in the logs might not be always correct (unless it's hardcoded somehow in amc connect)

ajanon commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,88 @@ jobs:
+        done
+        port=$(grep -oP 'adb connect 127.0.0.1:\K\d+' logs/anbox-connect.log)
+                  
+        adb connect 127.0.0.1:"$port"
```
Nitpick: why not get both IP and port here?

Untested, but something like:
```suggestion
        url=$(grep -oP 'adb connect\s*\K.*' logs/anbox-connect.log)
                  
        adb connect "${url}"
```

I'm not sure of the exact format of amc connect's output, so this might not work as-is.

Wdyt?

tarek-hanna replied:
sure i will try it but the original amc-connect already returns 127.0.0.1 with different port every run so i am not sure if this is necessary.

ajanon commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,88 @@ jobs:
+      if: always()
+      run: |
+        id=${{ steps.create-instance.outputs.id }}
+        # amc delete -y "$id"
```
Is that required?
If it is, why is it commented?

tarek-hanna replied:
if the instance is running on the remote amc which is the Customer's AWS .. we should clean it up, that' why i added the delete step. However if everything is running in the github hosted runner it will clean itself when the job is finished.

I commented it because i wanted to see the created instance on the remote aws.. 

anyway i reverted the comment

ajanon replied:
Ah, thanks for clarification, all good for me.

tarek-hanna commented:
@morphis @ajanon all the comments about the commented steps will be reverted as it takes too much time to run while the aws instance is running, which incurs running cost. so i disabled or commented longs steps ,, i will revert all of it and let you know

tarek-hanna committed:
feat: implement amc and adb connect on remote amc to run e2e test cases

This PR implements workflow step to "Setup remote amc" and then trust
the relevant certificates.

The remote amc is set to default and next step "Create Android Instance"
to launch an instance with streaming enabled

"Connect over adb" runs amc connect command to create a sharing link and
intiate adb connection as replacement for anbox-connect". Once the adb
url (ip & port) is available, adb connect is called to attach a device
in order to be able to execute the last step "Run e2e Tests" that
executes ./gradlew :app:connectedDemoDebugAndroidTest command

adglkh commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,100 @@ jobs:
       run: |
-        id="$(amc launch -r -s adb jammy:android${ANDROID_VERSION}:amd64)"
+        sudo snap restart anbox-cloud-appliance.ams
+        sleep 30
```
I’m not sure why we need to sleep for 30 seconds. If the delay is only due to the unix domain socket not being ready, a retry mechanism would be a more robust solution.

tarek-hanna replied:
sleep and restart removed

adglkh commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,100 @@ jobs:
+      if: always()
+      run: |
+        id=${{ steps.create-instance.outputs.id }}
+        amc delete -y "$id"
```
Maybe `amc delete -ya` for the simplicity

tarek-hanna replied:
i dont think we should delete all, if the "client" is running their own instances on the same remote. that's why we only delete the one we create for testing, and we def' shouldn't keep it unless we want to add a flag to retain the instances. WDYT?

adglkh commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,100 @@ jobs:
-        addr="$(amc show "$id" --format=json | jq -r .network.address)"
-        adb connect "$addr":5559
-        test "$(adb shell getprop ro.product.model)" = Anbox
+        rm -fr logs && mkdir -p logs
```
Are all logs under the logs folder collected if the final step of the workflow fails?

tarek-hanna replied:
no that's just a temp log to read the ip and port from `amc connect` output .. I can add a logging "step" and possibly include it in the artifact, but i suppose this is unnecessary at this point.

tarek-hanna replied:
logging step added :+1:

adglkh commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,100 @@ jobs:
       id: create-instance
       run: |
-        id="$(amc launch -r -s adb jammy:android${ANDROID_VERSION}:amd64)"
+        sudo snap restart anbox-cloud-appliance.ams
```
I don't get the point of why we need to restart ams service

tarek-hanna replied:
restart removed as it is no longer needed

adglkh commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,100 @@ jobs:
+        url=${REMOTE_AMC_URL}
+        if [ -z "${url}" ]; then
+            url="https://$(amc config show | grep -Po 'core\.https_address:\s*\K.*')"
+            amc config trust add "${dir}/client.crt"
```
Is there a guarantee that the client certificate was added to the trust store if the REMOTE_AMC_URL was given?

tarek-hanna replied:
For gh-runner the client.crt is already available. In case of Aws, the cert is manually trusted on the remote ams and provided in the secrets, to replace the runner's client.cert and client.key...in both scenarios client.crt is trusted.
ON the remote side, it is a pre-requisite for added security, and it should be clearly communicated when installing the appliance.

ajanon reviewed (approved):
LGTM

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,134 @@ jobs:
+        if [ -n "${REMOTE_SERVER_CERT}" ]; then 
+          mkdir -p "$HOME/snap/anbox-cloud-appliance/current/client/servercerts"
+          echo "${REMOTE_SERVER_CERT}" > "$HOME/snap/anbox-cloud-appliance/current/client/servercerts/server.crt"
+          amc config trust add "$HOME/snap/anbox-cloud-appliance/current/client/servercerts/server.crt"
```
This is not necessary. The certificate of the remote server is only needed on the client, not the server.

adglkh commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,134 @@ jobs:
+        sudo snap logs -n all anbox-cloud-appliance > appliance.log
+        sudo anbox-cloud-appliance.buginfo > appliance.buginfo
+        status=$(amc show "$id" --format=json | jq -r .status)
+          if [ "$status" = error ] ; then
```
Let's remove the redundant spaces

adglkh commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,134 @@ jobs:
+        if [ -n "${GH_RUNNER_CERT}" ] && [ -n "${GH_RUNNER_KEY}" ]; then
+          echo "${GH_RUNNER_CERT}" > runner.crt
+          echo "${GH_RUNNER_KEY}" > runner.key
+          cat runner.crt runner.key > runner.pem
```
You only need to add the certificate to the trust without touching the private key.

adglkh commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -21,44 +21,134 @@ jobs:
+        id=${{ steps.create-instance.outputs.id }}
+        if [ -n "$id" ]; then
+          amc delete -y "$id"
+        fi
```
As mentioned, for simplicity, the above command can be folded into a single line:
`amc delete -ya`

tarek-hanna replied:
@adglkh am not sure you've seen my reply regarding deletion but we shouldn't delete "--all" when we run on the client's remote appliance. in case they are running other instances that they do not wish to deleted unintentionally.   Or am i misunderstanding the -ya flag ?

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -13,52 +13,144 @@ jobs:
+            "$server_crt_path"
+        fi
+
+        # FIXME: workaround lp#2129004. Once fixed we can switch back to
```
The fix should be in 1.28/edge by now but I am fine to update this in a follow up PR.

tarek-hanna replied:
will include it now :)

tarek-hanna replied:
@morphis --channel=1.28/edge   &  `amc remote add server "${url}"` have not fixed the issue of fingerprint.
<img width="1204" height="480" alt="Screenshot from 2025-10-21 15-54-43" src="https://github.com/user-attachments/assets/21f86b32-50e4-4a08-bc42-b26b26fcca27" />

tarek-hanna replied:
i will mark this here as resolved and created a ticket to address it separately [use amc connect in nia-e2e workflow](https://warthogs.atlassian.net/browse/WD-30312)

morphis reviewed (approved):
LGTM, thanks for all the hard work!

tarek-hanna committed:
fix: update certificates handling in 'Setup Remote amc' and add logs dump

This commit fixes the certificate configuration, and adds logging.
Readme is updated with the new repo keys.

adglkh reviewed (approved):
LGTM
